nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/PutAzureCosmosDBRecord.java

```diff
@@ -0,0 +1,186 @@
+            }catch (ConflictException e) {
+                //  insert with unique id is expeected, but incase test data contains data with duplicate id(s).
+                this.container.upsertItem(record);
+            }
+        }
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        final FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        final ComponentLog logger = getLogger();
+        final RecordReaderFactory recordParserFactory = context.getProperty(RECORD_READER_FACTORY)
+                .asControllerService(RecordReaderFactory.class);
+
+        final String partitionKeyField = context.getProperty(PARTITION_KEY).getValue();
+        List<Map<String, Object>> batch = new ArrayList<>();
+        int ceiling = context.getProperty(INSERT_BATCH_SIZE).asInteger();
+        boolean error = false;
+        try (final InputStream inStream = session.read(flowFile);
+             final RecordReader reader = recordParserFactory.createRecordReader(flowFile, inStream, getLogger())) {
+
+            RecordSchema schema = reader.getSchema();
+            Record record;
+
+            while((record = reader.nextRecord()) != null) {
+                // Convert each Record to HashMap
+                Map<String, Object> contentMap = (Map<String, Object>) DataTypeUtils.convertRecordFieldtoObject(record, RecordFieldType.RECORD.getRecordDataType(schema));
+                if(contentMap.containsKey("id")) {
+                    Object val = contentMap.get("id");
+                    if(!(val instanceof String)) {
+                        logger.debug("coverting number id into string...");
+                        contentMap.put("id", contentMap.get("id").toString());
+                    }
+                } else {
+                    contentMap.put("id", flowFile.getAttribute("uuid"));
+                }
+                if(!contentMap.containsKey(partitionKeyField)){
+                    // We set partition key field value with flowfile.uuid if not exists
+                    contentMap.put(partitionKeyField, flowFile.getAttribute("uuid"));
+                }
+                batch.add(contentMap);
+                if (batch.size() == ceiling) {
+                    bulkInsert(batch);
+                    batch = new ArrayList<>();
+                }
+            }
+            if(batch.size() > 0) {
+                bulkInsert(batch);
+            }
+        } catch (SchemaNotFoundException | IOException | MalformedRecordException | CosmosClientException e) {
+            logger.error("PutAzureCosmoDBRecord failed with error:", e);
+            session.transfer(flowFile, REL_FAILURE);
+            error = true;
+        } finally {
+            if (!error) {
+                session.getProvenanceReporter().send(flowFile, getURI(context));
+                session.transfer(flowFile, REL_SUCCESS);
```

Comment:
Is there a reason this just isn't inside the `try() {}`?

Reply:
It seems that session.transfer should be called outside try catch block. When I moved them inside try and catch block, I am getting the following exception. I will keep those two lines in finally block as PutMongoRecord does, not inside try catch:


[ERROR] testOnTriggerWithFlatRecords(org.apache.nifi.processors.azure.cosmos.document.PutAzureCosmosDBRecordTest)  Time elapsed: 0.256 s  <<< FAILURE!
java.lang.AssertionError: java.lang.IllegalStateException: FlowFile[0,850021980779208.mockFlowFile,0B] already in use for an active callback or InputStream created by ProcessSession.read(FlowFile) has not been closed
        at org.apache.nifi.processors.azure.cosmos.document.PutAzureCosmosDBRecordTest.testOnTriggerWithFlatRecords(PutAzureCosmosDBRecordTest.java:128)
Caused by: java.lang.IllegalStateException: FlowFile[0,850021980779208.mockFlowFile,0B] already in use for an active callback or InputStream created by ProcessSession.read(FlowFile) has not been closed

[ERROR] testOnTriggerWithNestedRecords(org.apache.nifi.processors.azure.cosmos.document.PutAzureCosmosDBRecordTest)  Time elapsed: 0.013 s  <<< FAILURE!
java.lang.AssertionError: java.lang.IllegalStateException: FlowFile[0,850022330204874.mockFlowFile,0B] already in use for an active callback or InputStream created by ProcessSession.read(FlowFile) has not been closed
        at org.apache.nifi.processors.azure.cosmos.document.PutAzureCosmosDBRecordTest.testOnTriggerWithNestedRecords(PutAzureCosmosDBRecordTest.java:190)
Caused by: java.lang.IllegalStateException: FlowFile[0,850022330204874.mockFlowFile,0B] already in use for an active callback or InputStream created by ProcessSession.read(FlowFile) has not been closed